architecture/README.md

```diff
@@ -0,0 +1,190 @@
+    
+    fun performActionA() {
+        navigator.navigate(ActionA)
+    }   
+
+    fun performActionB() {
+        navigator.navigate(ActionB)
+    }
+
+}
+```
+
+On the platform side the created Navigator should specify the `NavigationSpec` to be used for navigating.
+Multiple specs exist per platform, including all common navigation patterns within the app (both new screens and nested elements) as well as navigating to common OS screens (e.g opening the mail app).
+
+```kotlin
+// Android
+val viewModel = SomeNavigatingViewModel(
+    Navigator { action ->
+        when (action) {
+            is ActionA -> NavigationSpec.Activity(SomeActivity::class.java)
+            is ActionB -> NavigationSpec.Fragment(R.id.some_fragment_container, createFragment = {SomeFragment()})
+        }   
+    })
+
+// iOS
+val viewModel = SomeNavigatingViewModeel(
+    Navigator(viewController) { action ->
+        when (action) {
+            is ActionA -> NavigationSpec.Present(present = { someViewController() })
+            is ActionB -> NavigationSpec.Nested(containerView = containerView, nested = {someNestedViewController()})
+        }
+    }
+)
+```
+
+To share data between navigating objects a `NavigationBundle` class can be passed via the `NavigationAction`.
+This Bundle supports all common data types as well as optionals, nested bundles and serializable objects.
+Values can be extracted from the bundle given a known `NavigationSpecRow`.
+Note that it is possible to request a NavigationSpecRow not supported by the Bundle, which will result in a `NavigationBundleGetError`.
+
+```kotlin
+sealed class SomeSpecRow<V>(associatedType: NavigationBundleSpecType<V>) : NavigationBundleSpecRow<V>(associatedType) {
+    object BooleanSpecRow : SomeSpecRow<Boolean>(NavigationBundleSpecType.BooleanType)
+    object SerializableSpecRow : SomeSpecRow<MockSerializable>(NavigationBundleSpecType.SerializedType(SomeSerializable.serializer()))
+    object OptionalString : SomeSpecRow<String?>(NavigationBundleSpecType.OptionalType(NavigationBundleSpecType.StringType))
+    object OptionalFloat : SomeSpecRow<Float?>(NavigationBundleSpecType.OptionalType(NavigationBundleSpecType.FloatType))
+}
+
+class SomeBundleSpec : NavigationBundleSpec<SomeSpecRow<*>>(setOf(SomeSpecRow.BooleanSpecRow, SomeSpecRow.SerializableSpecRow, SomeSpecRow.OptionalString, SomeSpecRow.OptionalFloat))
+
+sealed class SomeBundleNavigationAction<B: NavigationBundleSpecRow<*>>(bundle: NavigationBundle<B>) : NavigationAction<B>(bundle) {
+    class SomeAction(bundle: NavigationBundle<SomeSpecRow<*>>) : SomeNavigationAction<NavigationBundle<SomeSpecRow<*>>>(bundle)
+}
+
+class SomeBundleNavigatingViewModel(navigator: Navigator<SomeBundleNavigationAction<*>>): NavigatingViewModel<SomeNavigationAction>(navigator) {
+
+    fun performAction() {
+        navigator.navigate(SomeBundleNavigationAction.SomeAction(SomeBundleSpec().toBundle { row ->
```

Comment:
I made the suggestion to @Daeda88 to extract "bundle" into a seperate kaluga module. This could also provide the needed shared interface to build what you are asking.

Reply:
I _think_ this is also under the scope of #148, since it would have to solve this problem.